app/services/samples/metadata/update_service.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+module Samples
+  module Metadata
+    # Service used to Update Samples::Metadata
+    class UpdateService < BaseService
+      SampleMetadataUpdateError = Class.new(StandardError)
+      attr_accessor :sample, :metadata, :analysis_id
+
+      def initialize(project, sample, user = nil, params = {})
+        super(user, params)
+        @project = project
+        @sample = sample
+        @metadata = params['metadata']
+        @analysis_id = params['analysis_id']
+      end
+
+      def execute # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity, Metrics/MethodLength
+        authorize! sample.project, to: :update_sample?
+
+        if @project.id != @sample.project.id
+          raise SampleMetadataUpdateError,
+                I18n.t('services.samples.metadata.sample_does_not_belong_to_project', sample_name: @sample.name,
+                                                                                      project_name: @project.name)
+        end
+
+        if @metadata
+          # Without transforming keys, issues with overwritting can occur and multiples of the same key can appear
+          @metadata = @metadata.transform_keys(&:to_s)
+          metadata_fields_not_updated = []
+          @metadata.each do |key, value|
+            if value.blank?
+              if @sample['metadata'].key?(key)
+                @sample['metadata'].delete(key) && @sample['metadata_provenance'].delete(key)
+              end
+            else
+              metadata_updated = update_metadata(key, value)
+              !metadata_updated && metadata_fields_not_updated.append("#{key}: #{value}")
+            end
+          end
+          @sample.update(id: @sample.id)
+        else
+          raise SampleMetadataUpdateError,
```

Comment:
Just thinking about the next part of this which is going to be the metadata controller, we will need the service to return true/false if it was updated or not. The `@sample.update(id: @sample.id)` will return a boolean by default, but here we are just raising an error and not returning anything. I still need to test it out but I believe we can raise the error and do a return in this case we would return false after the error is raised. Then in the controller we can check if it was updated or not and if it was updated successfully we can do a secondary check to see if we need to return an error message for any keys that weren't updated

Reply:
Updated in 8a632ec0a9396186a5fa9bbadacc4b27e3f8c76b

I explicitly returned a `true` if `metadata_fields_not_updated.count == 0` as it seemed without it, the expected return was not behaving as expected. The problem I can see is if partial metadata fields were updated, and others not, in the same execution. 
I imagine in the view, we would want to provide both a success and failure feedback to the user via flash msg. Currently, by returning false if we have an error msg (ie: empty metadata, or partial metadata not updated), I don't see a clean way of still providing a success to the user that some fields were updated.

Would returning a list of updated metadata fields work? Return updated metadata fields to controller, and still check for errors in the controller.
If updated metadata count == 0 : Display error (whether it's empty metadata, or all fields not updated)
If updated metadata count positive? : still check error and run both success and error if error (captures both all metadata updated, or partial updated/not updated)